README.md
# Flarum Laravel Session <!-- omit in toc -->

![tests](https://github.com/lbausch/flarum-laravel-session/workflows/tests/badge.svg) [![codecov](https://codecov.io/gh/lbausch/flarum-laravel-session/branch/master/graph/badge.svg)](https://codecov.io/gh/lbausch/flarum-laravel-session)

- [What It Does](#what-it-does)
- [Requirements](#requirements)
- [Installation and Configuration](#installation-and-configuration)
  - [Composer](#composer)
  - [Register Middleware](#register-middleware)
  - [Setup Database Connection](#setup-database-connection)
  - [Publish Package Configuration](#publish-package-configuration)
  - [Disable Cookie Encryption](#disable-cookie-encryption)
- [Usage](#usage)
  - [Setup Middleware](#setup-middleware)
  - [Handle An Identified User](#handle-an-identified-user)
- [Accessing Flarum Session Cookie From Different Domain](#accessing-flarum-session-cookie-from-different-domain)

## What It Does
This package allows to use the session of [Flarum](https://flarum.org/) for authentication within a Laravel application.
It accesses Flarum's session cookie and reads the session data from the session storage.
Based on the user information in the Flarum user database an user in the Laravel application is created / updated and logged in.

## Requirements
+ PHP 7.3+
+ Laravel 8
+ Working installation of Flarum in the same filesystem as the Laravel application, so Flarum's session files can be read
+ Flarum and Laravel need to share the same domain / subdomain, so Flarum's session cookie can be accessed

## Installation and Configuration

### Composer
Install the package with Composer:
```bash
composer require lbausch/flarum-laravel-session
```

### Register Middleware
Register the `\Bausch\FlarumLaravelSession\FlarumSessionMiddleware` middleware in `app/Http/Kernel.php`:
```php
/**
 * The application's route middleware.
 *
 * These middleware may be assigned to groups or used individually.
 *
 * @var array
 */
protected $routeMiddleware = [
    // ...
    'flarum' => \Bausch\FlarumLaravelSession\FlarumSessionMiddleware::class,
    // ...
];
```

### Setup Database Connection
Define a database connection for the Flarum database in `config/database.php`:
```php
'flarum' => [
    'driver' => 'mysql',
    'url' => env('FLARUM_DATABASE_URL'),
    'host' => env('FLARUM_DB_HOST', '127.0.0.1'),
    'port' => env('FLARUM_DB_PORT', '3306'),
    'database' => env('FLARUM_DB_DATABASE', 'forge'),
    'username' => env('FLARUM_DB_USERNAME', 'forge'),
    'password' => env('FLARUM_DB_PASSWORD', ''),
    'unix_socket' => env('FLARUM_DB_SOCKET', ''),
    'charset' => 'utf8mb4',
    'collation' => 'utf8mb4_unicode_ci',
    'prefix' => '',
    'prefix_indexes' => true,
    'strict' => true,
    'engine' => null,
    'options' => extension_loaded('pdo_mysql') ? array_filter([
        PDO::MYSQL_ATTR_SSL_CA => env('FLARUM_MYSQL_ATTR_SSL_CA'),
    ]) : [],
],
```

### Publish Package Configuration
Publish the package configuration using `php artisan vendor:publish --provider=Bausch\\FlarumLaravelSession\\ServiceProvider` and update `config/flarum.php` with your settings.

### Disable Cookie Encryption
To avoid Laravel from trying to encrypt the Flarum session cookie, add the following to `app/Http/Middleware/EncryptCookies.php`:
```php
/**
 * The names of the cookies that should not be encrypted.
 *
 * @var array
 */
protected $except = [
    'flarum_session',
];
```

## Usage

### Setup Middleware
In `routes/web.php` you may assign the middleware as desired:
```php
Route::middleware(['flarum'])->group(function () {
    Route::get('/', function () {
        return view('welcome');
    });
});
```
All requests to the `/` route will then be checked by the middleware.

### Handle An Identified User
Once the middleware successfully identified an user, it executes the default handler `\Bausch\FlarumLaravelSession\Actions\HandleIdentifiedUser`. You may configure a different handler by calling `FlarumLaravelSession::handleIdentifiedUser()` in a service provider. This is a perfect place to update attributes or execute further actions, just remember to implement the `\Bausch\FlarumLaravelSession\Contracts\FlarumUserIdentified` interface.
Have a look at the default handler for a reference implementation.

```php
<?php

namespace App\Providers;

use Illuminate\Support\ServiceProvider;
use Bausch\FlarumLaravelSession\FlarumLaravelSession;
use App\Handlers\YourCustomHandler;

class AppServiceProvider extends ServiceProvider
{
    /**
     * Register any application services.
     *
     * @return void
     */
    public function register()
    {
        //
    }

    /**
     * Bootstrap any application services.
     *
     * @return void
     */
    public function boot()
    {
        FlarumLaravelSession::handleIdentifiedUser(YourCustomHandler::class);
    }
}
```

If you need to use a different user model than `App\Models\User`, you may call `FlarumLaravelSession::useUserModel(YourUser::class)` in your service provider.

## Accessing Flarum Session Cookie From Different Domain
If Flarum is running on domain.tld and Laravel on sub.domain.tld you need to configure Flarum (`config.php`), so the session cookie can be accessed on the subdomain:
```php
// Note the dot before domain.tld
'cookie' => ['domain' => '.domain.tld'],
```


CHANGELOG.md
# Changelog

Please see [Releases](https://github.com/lbausch/flarum-laravel-session/releases).


tests/TestCase.php
<?php

namespace Tests;

use Bausch\FlarumLaravelSession\Actions\HandleIdentifiedUser;
use Bausch\FlarumLaravelSession\FlarumLaravelSession;
use Bausch\FlarumLaravelSession\ServiceProvider;
use Orchestra\Testbench\TestCase as BaseTestCase;
use Tests\Models\User;

abstract class TestCase extends BaseTestCase
{
    /**
     * Setup the test environment.
     */
    protected function setUp(): void
    {
        parent::setUp();

        $this->loadMigrationsFrom(__DIR__.'/database/migrations');

        $this->artisan('migrate', ['--database' => 'testing']);
    }

    /**
     * Define environment setup.
     *
     * @param \Illuminate\Foundation\Application $app
     *
     * @return void
     */
    protected function getEnvironmentSetUp($app)
    {
        // Flarum configuration
        $app['config']->set('flarum', [
            'url' => 'https://flarum.tld',
            'session' => [
                'cookie' => 'flarum_session',
                'path' => 'storage',
            ],
            'db_connection' => 'flarum',
        ]);

        // Local database
        $app['config']->set('database.default', 'testing');
        $app['config']->set('database.connections.testing', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);

        // Flarum database
        $app['config']->set('database.connections.flarum', [
            'driver' => 'sqlite',
            'url' => null,
            'database' => __DIR__.'/database/flarum.sqlite',
            'prefix' => '',
            'foreign_key_constraints' => true,
        ]);

        FlarumLaravelSession::handleIdentifiedUser(HandleIdentifiedUser::class);
        FlarumLaravelSession::useUserModel(User::class);
    }

    /**
     * Load Service Provider.
     *
     * @param \Illuminate\Foundation\Application $app
     *
     * @return array
     */
    protected function getPackageProviders($app)
    {
        return [
            ServiceProvider::class,
        ];
    }
}


tests/Models/User.php
<?php

namespace Tests\Models;

use Illuminate\Foundation\Auth\User as Authenticatable;

class User extends Authenticatable
{
}


tests/FlarumMiddlewareTest.php
<?php

declare(strict_types=1);

namespace Tests;

use Bausch\FlarumLaravelSession\FlarumSessionMiddleware;
use Illuminate\Filesystem\Filesystem;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Config;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Str;
use Symfony\Component\HttpKernel\Exception\HttpException;

final class MiddlewareTest extends TestCase
{
    public function testExceptionIsThrownIfSessionCouldNotBeFound()
    {
        $exception_was_thrown = false;

        try {
            $response = (new FlarumSessionMiddleware())->handle(new Request(), function ($request) {});
        } catch (HttpException $exception) {
            $this->assertEquals(403, $exception->getStatusCode());

            $exception_was_thrown = true;
        }

        $this->assertTrue($exception_was_thrown);
    }

    public function testInvalidSessionCookieResultsInRedirectToFlarum()
    {
        $request = new Request($query = [], $request = [], $attributes = [], $cookies = [
            'flarum_session' => 'foobar',
        ]);

        $response = (new FlarumSessionMiddleware())->handle($request, function ($request) {});

        $this->assertEquals(Config::get('flarum.url'), $response->getTargetUrl());
    }

    public function testMiddlewareExecutesNextRequest()
    {
        $session_id = $this->generateSessionId();
        $session_file = Config::get('flarum.session.path').'/'.$session_id;

        // Mock Filesystem
        $this->partialMock(Filesystem::class, function ($mock) use ($session_file) {
            $mock->shouldReceive('isFile')
                ->once()
                ->with($session_file)
                ->andReturn(true);

            $mock->shouldReceive('lastModified')
                ->once()
                ->with($session_file)
                ->andReturn(time());

            $mock->shouldReceive('sharedGet')
                ->once()
                ->with($session_file)
                ->andReturn(serialize(['access_token' => 'foobar']));
        });

        $request = new Request($query = [], $request = [], $attributes = [], $cookies = [
            'flarum_session' => $session_id,
        ]);

        $next_request_executed = false;

        $response = (new FlarumSessionMiddleware())->handle($request, function ($request) use (&$next_request_executed) {
            $next_request_executed = true;
        });

        $this->assertTrue($next_request_executed);

        // Verify user was created
        $users = DB::table('users')->get();

        $this->assertEquals(1, $users->count());

        $user = $users->first();

        $this->assertEquals(1, $user->flarum_id);
        $this->assertEquals('testuser', $user->username);
        $this->assertEquals('testuser@flarum.tld', $user->email);
    }

    public function testUsersDontGetCreatedTwice()
    {
        $session_id = $this->generateSessionId();
        $session_file = Config::get('flarum.session.path').'/'.$session_id;

        // Mock Filesystem
        $this->partialMock(Filesystem::class, function ($mock) use ($session_file) {
            $mock->shouldReceive('isFile')
                ->once()
                ->with($session_file)
                ->andReturn(true);

            $mock->shouldReceive('lastModified')
                ->once()
                ->with($session_file)
                ->andReturn(time());

            $mock->shouldReceive('sharedGet')
                ->once()
                ->with($session_file)
                ->andReturn(serialize(['access_token' => 'foobar']));
        });

        $request = new Request($query = [], $request = [], $attributes = [], $cookies = [
            'flarum_session' => $session_id,
        ]);

        $next_request_executed = false;

        $response = (new FlarumSessionMiddleware())->handle($request, function ($request) use (&$next_request_executed) {
            $next_request_executed = true;
        });

        $this->assertTrue($next_request_executed);

        $users = DB::table('users')->get();

        $this->assertEquals(1, $users->count());

        $user = $users->first();

        $this->assertEquals(1, $user->flarum_id);
        $this->assertEquals('testuser', $user->username);
        $this->assertEquals('testuser@flarum.tld', $user->email);
    }

    public function testMiddlewareExecutesNextRequestEarly()
    {
        $session_id = $this->generateSessionId();
        $session_file = Config::get('flarum.session.path').'/'.$session_id;

        // Mock Filesystem
        $this->partialMock(Filesystem::class, function ($mock) use ($session_file) {
            $mock->shouldReceive('isFile')
                ->once()
                ->with($session_file)
                ->andReturn(true);

            $mock->shouldReceive('lastModified')
                ->once()
                ->with($session_file)
                ->andReturn(time());

            $mock->shouldReceive('sharedGet')
                ->once()
                ->with($session_file)
                ->andReturn(serialize(['access_token' => 'foobar']));
        });

        $request = new Request($query = [], $request = [], $attributes = [], $cookies = [
            'flarum_session' => $session_id,
        ]);

        $next_request_executed = false;

        $response = (new FlarumSessionMiddleware())->handle($request, function ($request) use (&$next_request_executed) {
            $next_request_executed = true;
        });

        $this->assertTrue($next_request_executed);

        // Mock Filesystem
        $this->partialMock(Filesystem::class, function ($mock) {
            $mock->shouldReceive('isFile')
                ->never();

            $mock->shouldReceive('lastModified')
                ->never();

            $mock->shouldReceive('sharedGet')
                ->never();
        });

        $next_request_executed = false;

        $response = (new FlarumSessionMiddleware())->handle($request, function ($request) use (&$next_request_executed) {
            $next_request_executed = true;
        });

        $this->assertTrue($next_request_executed);
    }

    public function testMiddlewareRedirectsIfUserWasNotFound()
    {
        $session_id = $this->generateSessionId();
        $session_file = Config::get('flarum.session.path').'/'.$session_id;

        // Mock Filesystem
        $this->partialMock(Filesystem::class, function ($mock) use ($session_file) {
            $mock->shouldReceive('isFile')
                ->once()
                ->with($session_file)
                ->andReturn(true);

            $mock->shouldReceive('lastModified')
                ->once()
                ->with($session_file)
                ->andReturn(time());

            $mock->shouldReceive('sharedGet')
                ->once()
                ->with($session_file)
                ->andReturn(serialize(['access_token' => 'loremipsum']));
        });

        $request = new Request($query = [], $request = [], $attributes = [], $cookies = [
            'flarum_session' => $session_id,
        ]);

        $exception_was_thrown = false;

        try {
            $response = (new FlarumSessionMiddleware())->handle($request, function ($request) {});
        } catch (HttpException $exception) {
            $this->assertEquals(403, $exception->getStatusCode());

            $exception_was_thrown = true;
        }

        $this->assertTrue($exception_was_thrown);
    }

    /**
     * Generate session id.
     */
    protected function generateSessionId(): string
    {
        return Str::random(40);
    }
}


src/FlarumMiddleware.php
<?php

namespace Bausch\FlarumLaravelSession;

use Closure;
use Illuminate\Container\Container;
use Illuminate\Filesystem\Filesystem;
use Illuminate\Foundation\Auth\User;
use Illuminate\Http\Request;
use Illuminate\Session\FileSessionHandler;
use Illuminate\Session\Store;
use Illuminate\Support\Facades\Auth;
use Illuminate\Support\Facades\Config;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Str;

class FlarumMiddleware
{
    /**
     * Handle an incoming request.
     *
     * @return mixed
     */
    public function handle(Request $request, Closure $next)
    {
        // If the user is already authenticated, continue
        if (Auth::check()) {
            // dd('asf');
            return $next($request);
        }

        // Get content of Flarum session cookie
        $flarum_session_id = $request->cookie(Config::get('flarum.session.cookie', false));

        // Abort if no session id could be found
        if (!$flarum_session_id) {
            abort(403);
        }

        // Create session store
        $session_store = new Store('flarum-laravel-session', $this->getFileSessionHandler(), $flarum_session_id);

        // Start session
        $session_store->start();

        // Try to get user id
        $user_id = $session_store->get('user_id', false);

        // Redirect if no user id was found
        if (!$user_id) {
            return redirect(Config::get('flarum.url'));
        }

        // Get Flarum user from Flarum database
        $flarum_user = DB::connection(Config::get('flarum.db_connection'))->table('users')->find($user_id);

        // Abort if no Flarum user is present in database
        if (!$flarum_user) {
            abort(403);
        }

        // Find the corresponding local user
        $user = $this->getUser()->where('flarum_id', $flarum_user->id)->first();

        // Create or update the corresponding local user
        $user = $this->createOrUpdateUser($user, $flarum_user);

        // Login the local user and remember him
        Auth::login($user, true);

        return $next($request);
    }

    /**
     * Get FileSessionHandler.
     */
    protected function getFileSessionHandler(): FileSessionHandler
    {
        // Create filesystem
        $filesystem = Container::getInstance()->make(Filesystem::class);

        // Get path to session files
        $session_path = Config::get('flarum.session.path');

        // Session lifetime in minutes
        $lifetime_minutes = 120;

        return new FileSessionHandler(
            $filesystem,
            $session_path,
            $lifetime_minutes
        );
    }

    /**
     * Create or update User.
     */
    protected function createOrUpdateUser(?User $user, object $flarum_user): User
    {
        // Get a user instance
        if (null === $user) {
            $user = $this->getUser();
        }

        // Attributes to update: Flarum user => local user
        $update_attributes = Config::get('flarum.update_attributes', []);

        // Update attributes
        foreach ($update_attributes as $flarum_attribute => $local_attribute) {
            $user->{$local_attribute} = $flarum_user->{$flarum_attribute};
        }

        // Set a random password
        $user->password = bcrypt(Str::random(30));

        // Save user
        if ($user->isDirty()) {
            $user->save();
        }

        // Return user
        return $user;
    }

    /**
     * Get User instance.
     */
    protected function getUser(): User
    {
        return Container::getInstance()->make(Config::get('flarum.model'));
    }
}


src/FlarumLaravelSession.php
<?php

namespace Bausch\FlarumLaravelSession;

use Bausch\FlarumLaravelSession\Contracts\FlarumUserIdentified;

class FlarumLaravelSession
{
    /**
     * The user model that should be used by FlarumLaravelSession.
     *
     * @var string
     */
    public static $userModel = 'App\\Models\\User';

    /**
     * Register class that handles actions once a Flarum user has been identified.
     */
    public static function handleIdentifiedUser(string $class)
    {
        return app()->singleton(FlarumUserIdentified::class, $class);
    }

    /**
     * Get the name of the user model used by the application.
     *
     * @return string
     */
    public static function userModel()
    {
        return static::$userModel;
    }

    /**
     * Specify the user model that should be used by FlarumLaravelSession.
     *
     * @return static
     */
    public static function useUserModel(string $model)
    {
        static::$userModel = $model;

        return new static();
    }
}


src/Actions/HandleIdentifiedUser.php
<?php

namespace Bausch\FlarumLaravelSession\Actions;

use Bausch\FlarumLaravelSession\Contracts\FlarumUserIdentified;
use Bausch\FlarumLaravelSession\FlarumLaravelSession;
use Closure;
use Illuminate\Container\Container;
use Illuminate\Foundation\Auth\User;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Auth;
use Illuminate\Support\Str;

class HandleIdentifiedUser implements FlarumUserIdentified
{
    public function __invoke(object $flarum_user, Request $request, Closure $next)
    {
        // Find the corresponding local user
        $user = $this->getUser()->where('flarum_id', $flarum_user->id)->first();

        // Create or update the corresponding local user
        $user = $this->createOrUpdateUser($user, $flarum_user);

        // Login the local user and remember him
        Auth::login($user, $remember = true);

        return $next($request);
    }

    /**
     * Create or update User.
     *
     * @param \Illuminate\Foundation\Auth\User $user
     */
    protected function createOrUpdateUser(?User $user, object $flarum_user): User
    {
        // Get a user instance
        if (null === $user) {
            $user = $this->getUser();
        }

        $user = $this->updateAttributes($user, $flarum_user);

        // Save user
        if ($user->isDirty()) {
            $user->save();
        }

        // Return user
        return $user;
    }

    /**
     * Update attributes.
     *
     * @param \Illuminate\Foundation\Auth\User $user
     */
    protected function updateAttributes(?User $user, object $flarum_user): User
    {
        // Attributes to update: Flarum user => local user
        $update_attributes = [
            'username' => 'username',
            'id' => 'flarum_id',
            'email' => 'email',
        ];

        // Update attributes
        foreach ($update_attributes as $flarum_attribute => $local_attribute) {
            $user->{$local_attribute} = $flarum_user->{$flarum_attribute};
        }

        if ($user->isDirty()) {
            // Set a random password
            $user->password = bcrypt(Str::random(30));
        }

        return $user;
    }

    /**
     * Get user instance.
     */
    protected function getUser(): User
    {
        return Container::getInstance()->make(FlarumLaravelSession::userModel());
    }
}


src/FlarumSessionMiddleware.php
<?php

namespace Bausch\FlarumLaravelSession;

use Bausch\FlarumLaravelSession\Contracts\FlarumUserIdentified;
use Closure;
use Illuminate\Container\Container;
use Illuminate\Filesystem\Filesystem;
use Illuminate\Http\Request;
use Illuminate\Session\FileSessionHandler;
use Illuminate\Session\Store;
use Illuminate\Support\Facades\Auth;
use Illuminate\Support\Facades\Config;
use Illuminate\Support\Facades\DB;

class FlarumSessionMiddleware
{
    /**
     * Handle an incoming request.
     *
     * @return mixed
     */
    public function handle(Request $request, Closure $next)
    {
        // If the user is already authenticated, continue
        if (Auth::check()) {
            return $next($request);
        }

        // Get content of Flarum session cookie
        $flarum_session_id = $request->cookie(Config::get('flarum.session.cookie', false));

        // Abort if no session id could be found
        if (!$flarum_session_id) {
            abort(403);
        }

        // Create session store
        $session_store = new Store('flarum-laravel-session', $this->getFileSessionHandler(), $flarum_session_id);

        // Start session
        $session_store->start();

        // Try to get access_token
        $access_token = $session_store->get('access_token', false);

        // Redirect if no access_token was found
        if (!$access_token) {
            return redirect(Config::get('flarum.url'));
        }

        // Try to get token data with user id
        $token_data = $flarum_user = DB::connection(Config::get('flarum.db_connection'))->table('access_tokens')->where('token', $access_token)->first();

        // Redirect if no token was found
        if (!$token_data || !$token_data->user_id) {
            return redirect(Config::get('flarum.url'));
        }

        $user_id = $token_data->user_id;

        // Get Flarum user from Flarum database
        $flarum_user = DB::connection(Config::get('flarum.db_connection'))->table('users')->find($user_id);

        // Abort if no Flarum user is present in database
        if (!$flarum_user) {
            abort(403);
        }

        // Get Handler for handling Flarum user
        $handler = Container::getInstance()->make(FlarumUserIdentified::class);

        // Execute handler
        return $handler($flarum_user, $request, $next);
    }

    /**
     * Get FileSessionHandler.
     */
    protected function getFileSessionHandler(): FileSessionHandler
    {
        // Create filesystem
        $filesystem = Container::getInstance()->make(Filesystem::class);

        // Get path to session files
        $session_path = Config::get('flarum.session.path');

        // Session lifetime in minutes
        $lifetime_minutes = 120;

        return new FileSessionHandler(
            $filesystem,
            $session_path,
            $lifetime_minutes
        );
    }
}
